### PR TITLE
Improve gracefully filling warning

### PR DESCRIPTION
**Pull request by znicholls**
This PR changes the behaviour of `iris.load` so that it only issues warnings about gracefully filling data and bounds if it actually does so.

### CODE REVIEW

**Comment by dkillick** [comment]
Hi @znicholls – congratulations on your first Iris PRs, and thanks for taking the time to make the submission! There are just a couple of things we could do with from you...

* you'll see that one of the checks that's failing is the CLA (Contributor License Agreement) checker. We ask all contributors to Iris (and the parent "SciTools" organisation) to submit a CLA with their first PR. Details on this can be found at https://scitools.org.uk/organisation.html#governance, in the "Contributors" bullet point.
* It looks like you've pulled in the docs changes you submitted in #3170 into this PR. Would you mind backing those changes out? No reason to submit the same things twice! It'll also make the changes you're proposing in this PR easier to spot.

Thanks again!

**Comment by znicholls** [comment]
thanks @dkillick. I've just sent the CLA through as advised on the governance page. 

I've just backed out the changes. My editor is setup to strip whitespace at the end of lines by default, if that's a problem I can remove all those changes too.

**Comment by dkillick** [comment]
Typically we try to avoid mixing code changes and stylistic changes in the same commit or PR, because it can make it harder to follow what's going on in the PR. This can, however, lead to stylistic changes never getting done! (As seems to be the case with this file). So not to worry about the whitespace removal; and thanks in fact for tidying up that file!

We'll try and progress the CLA soon as well - thanks for submitting one.

File: lib/iris/tests/unit/fileformats/pyke_rules/compiled_krb/fc_rules_cf_fc/test_build_dimension_coordinate.py [hunk 1]
```diff
@@ -112,6 +112,60 @@ def test_dim_coord_construction(self):
...
+            # loading.
+            with self.deferred_load_patch, self.get_cf_bounds_var_patch:
+                build_dimension_coordinate(self.engine, self.cf_coord_var)
+
+                # Test that expected coord is built and added to cube.
+                self.engine.cube.add_dim_coord.assert_called_with(
+                    expected_coord, [0])
+
+            # Assert no warning is raised
+            assert len(w) == 0
```
Comment (dkillick) [inline lib/iris/tests/unit/fileformats/pyke_rules/compiled_krb/fc_rules_cf_fc/test_build_dimension_coordinate.py]: These new tests look great – but would you mind adding testing that covers the cases where the incoming coordinate has masked **bounds**? Thanks!

File: lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb [hunk 2]
```diff
@@ -1621,7 +1621,7 @@ fc_extras
         attr_units = get_attr_units(cf_coord_var, attributes)
         points_data = cf_coord_var[:]
         # Gracefully fill points masked array.
-        if ma.isMaskedArray(points_data):
+        if (ma.isMaskedArray(points_data)) and (not (points_data.mask == False).all()):
```
Comment (dkillick) [inline lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb]: You know, I think you can achieve equivalent behaviour with `if ma.is_masked(points_data)`. Would you mind updating this line to use that masked array check instead?

File: lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb [hunk 3]
```diff
@@ -1631,7 +1631,7 @@ fc_extras
         if cf_bounds_var is not None:
             bounds_data = cf_bounds_var[:]
             # Gracefully fill bounds masked array.
-            if ma.isMaskedArray(bounds_data):
+            if (ma.isMaskedArray(bounds_data)) and (not (bounds_data.mask == False).all()):
```
Comment (dkillick) [inline lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb]: And again with this line.

File: lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb (hunk 2 above)
Comment (znicholls) [reply lib/iris/fileformats/_pyke_rules/fc_rules_cf.krb]: super nice pick up

**Comment by znicholls** [comment]
> Typically we try to avoid mixing code changes and stylistic changes in the same commit or PR, because it can make it harder to follow what's going on in the PR.

Makes sense. To get the tests passing locally I had to update a bunch of headers. I can remove those changes again if you'd like but the tests may fail?

File: lib/iris/tests/unit/fileformats/pyke_rules/compiled_krb/fc_rules_cf_fc/test_build_dimension_coordinate.py (hunk 1 above)
Comment (znicholls) [reply lib/iris/tests/unit/fileformats/pyke_rules/compiled_krb/fc_rules_cf_fc/test_build_dimension_coordinate.py]: have a look at the latest commit, I think this is what you wanted?

**Comment by dkillick** [comment]
> I had to update a bunch of headers

Yes that does happen sometimes. Don't worry about it!

There are also a bunch of Python 3 test failures. It very much looks like these aren't being caused by this PR, so (separately from this PR) some work will need to happen to get those tests all passing again...

**Comment by dkillick** [comment]
> a bunch of Python 3 test failures

I think these should be fixed by #3175 being merged. I'm going to restart the tests and see where that leaves us (now that the CLA checker bot is happy!).

**Comment by dkillick** [comment]
Congratulations @znicholls for getting your first Iris PR merged 🎉 The change here is much appreciated!

**Comment by znicholls** [comment]
Thanks @dkillick :) Shall I close #3170 as I think it's redundant?

**Comment by dkillick** [comment]
>Shall I close #3170 ... ?

Actually, I think the note on [how to run the tests](https://github.com/SciTools/iris/blob/b10861647963c3e818b6aee4c1e4c1896f1c90b5/INSTALL#L69-L82) in #3170 is quite useful and not replicated here, so I think there's still value to it.

**Pull request merged**